File: coinChange2/coin2.py
class Solution(object):
    #this solution usiing 2d array
    def change(self, amount, coins):
        if amount == 0:
            return 1
        if(len(coins) == 0) or amount < coins[0]:
            return 0
        dpList = [([1]+[0]*(amount)) for i in range(len(coins))]
        for i in range(len(coins)):
            for j in range(1,amount+1):
                dpList[i][j] = (dpList[i-1][j] if i-1 >= 0 else 0)+ (dpList[i][j-coins[i]] if j-coins[i] >= 0 else 0) 
        
        return dpList[-1][-1]
       

File: Coin_Change_2/coin2.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
       if (amount == 0)
            return 1;
        if(coins.size() == 0 || amount < coins[0])
            return 0;
        
        vector<int> arr(amount+1,0);
        arr[0]=1;
        for (int coin : coins)
            for (int i = 1 ; i <amount+1; i++)
                arr[i] += (i-coin >=0)? arr[i-coin]:0;
        
        return arr[amount];
        
    }
};
